### PR TITLE
Logbook generator for ILL data

### LINKED ISSUE #30244: Logbook generator for ILL experiments
Description
-----------

New algorithm is needed that will crawl through a given rawdata directory, read a predefined nexus entries without loading the files, and make an index of runs (logbook) keyed on unique run number. This is required for practically all the techniques. The list of relevant metadata however depends on the technique. List of pre-defined metadata could be put in instrument parameters, but there should be a way also to add new metadata with the algorithm.

Inputs
------

Raw data directory [mandatory]

Numor range [optional]

Facility [selection based on Facilities.xml, Default ILL]

Instrument [selection based on the chosen facility]

OutputWorkspace [mandatory table workspace]

OutputFile [optional excel readable CSV with .lbook]

MetadataEntries [, separated nexus paths for additional metadata to include in the table]

MetadataHeaders [, separated names of those additional entries, need to match the entries in size]

Language: Python with h5py

Note, this should not load the files, but only the corresponding entries with h5py, and the entries must be nexus paths (i.e. with /) and not sample logs.

### PR DESCRIPTION
**Pull request by darominski**
**Description of work.**

Generates logbook (`TableWorkspace`) containing all default logbook entries and headers that are pre-defined in the IPF relevant to the chosen instrument. Currently supported instruments: D11, D11B, D16, D22, D22B, D33, IN4, IN5, IN6, PANTHER, D7 (list of parameters not finalized) form the ILL.

There are three levels of the entries: default, optional and custom. Default entries are always read and included in the logbook. Optional entries are pre-defined in the relevant IPF and are included in the logbook if a header for the entry is specified. The last one is added when the full path in the NeXus is provided as a comma separated list.

Basic binary arithmetic operations: addition (the only one supported if the entry data is a string), subtraction, multiplication and division on entries data are supported. At the moment, no support is offered for organizing the operations using parentheses or scaling by read data by constant numbers.

**To test:**

1. From the workbench, specify ILL/D7 UnitTest data directory as 'Directory', and a numor range "394458, 394882". Giving no numor range will result in some rows being filled with "Not found" messages, as this directory contains also processed data that has different structure than raw data. Setting 'D7' as the instrument is mandatory.
2. Inspect the created `TableWorkspace`:
a) How many rows does it have? It should be exactly the number of files contained in the directory, unless the numor range has been speciifed and excludes these files. In the provided example range, there should be 2 rows.
b) What data is contained in the table? If the OptionalHeaders were not specified, it should contain only the entries specified in the IPF, currently: 'run_number', 'wavelength',  'experiment_identifier', 'start_time', 'end_time', 'duration'. Do the values make sense?
3. Specify a path and a name for the `OutputFile` property. Check if the created .csv file can be imported by your preferred spreadsheet software, e.g. LibreOffice Calc or Microsoft Excel.
3. Try setting a different instrument:
a) One that does not belong to the ILL should not go through validation, 
b) One that is not currently supported should raise an error that no default parameters are specified,
c) One that is supported but is not D7: then the output should contain many fields filled with "Not found" message, for every entry that was not found in the NeXus files. 

Fixes #30244 . 

Release notes added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py
Comment (DannyHindson) [inline Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py:65]: I'm not sure this test actually triggers the use of a binary operation (+). It looks like you've either got to put the formula for polarisation in here:
```
CustomEntries='/entry0/acquisition_mode, /entry0/D7/POL/actual_state+/entry0/D7/POL/actual_stateB1B2'
```
or you need to specify polarisation in the OptionalHeaders parameter. Please can you make one of these changes so the test executes the binary operation code in GenerateLogbook.py?

If do the first approach then it seems like the code to determine the column header in the output table workspace doesn't work 100% correctly in that it decides the column is called actual_stateB1B2 (just strips off the final string). This point about the column header not a big deal though.

File: docs/source/algorithms/GenerateLogbook-v1.rst
Comment (DannyHindson) [inline docs/source/algorithms/GenerateLogbook-v1.rst:59]: Please can you add a second usage example showing an example of the binary operation? The polarisation example from the unit test would be fine

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've been through and added a few details level comments. I've got one general question as well. What's the reason for all of the reference file changes? At first I thought it was because you'd added a section to the instrument parameter files and this was causing some changes in the instrument definition to show up in the system test comparisons between the generated \reference files. But it looks like you've disabled the instrument check in the system tests?

**Comment by gvardany** [comment]
> I've been through and added a few details level comments. I've got one general question as well. What's the reason for all of the reference file changes? At first I thought it was because you'd added a section to the instrument parameter files and this was causing some changes in the instrument definition to show up in the system test comparisons between the generated \reference files. But it looks like you've disabled the instrument check in the system tests?

That's a good question. The references should not be changed, but the instrument check should be disabled in validation, since it is a lot work to update all for simple instrument parameter changes like this.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:110]: Yes, of course! That will be corrected.

File: instrument/D11B_Parameters.xml
Comment (darominski) [reply instrument/D11B_Parameters.xml:29]: I agree this is not a pretty solution, intended for readability and ease of future modifications. I agree that your solution does look better and XML-like (but one should use slashes instead of backslashes), however it doesn't seem to work with Mantid. Unfortunately, only the final defined value is registered, and the rest is lost.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:78]: I indeed have come across MFP multiple times, and I am not convinced it would be the best choice here. First, it would start searching for the provided numors in the user search directories, not only in the desired directory which can cause confusion unless it is specified as an alternative workflow for this algorithm. Second, the validation of large number of files may be slow, as suggested by issue #29881. Also, the very convenient, in other circumstances, workspace operations possible with MFP would need to be disallowed. 

I am going to check how fast it can deal with a couple of thousand of files on a remote drive, and if it slows down the execution I would rather not use it.

File: Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py
Comment (darominski) [reply Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py:65]: Yes indeed, this is not correct. The OptionalHeaders need to be defined for this operation to be performed. The original operation would search for a single entry of the provided structure, which was not possible to be found. This test is fixed and also the name creation has been updated to be derived from the entries going into the operation to be performed.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (DannyHindson) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:78]: If you've noted that the MFP type exists and have decide not to use it that's fine - just wanted to double check. If you decide not to use it then I still think it's worth using : for the range so people used to specifying ranges in the MFP can use the same syntax here for a range

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:44]: Great, thanks. Somehow, I wasn't aware of this validator.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:294]: Binary operations still work if the inputs contain np.array or ndarray. I shall add a special case to improve the formatting of the output in that case, as you suggested.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py]: I replaced `IntArrayOrderedPairsValidator` with `IntArrayBoundedValidator`, and with a couple of lines of custom validation it provides the desired functionality, where the user can specify a numor range using a colon or a dash, along with giving two boundary values.

**Comment by darominski** [comment]
> I've been through and added a few details level comments. I've got one general question as well. What's the reason for all of the reference file changes? At first I thought it was because you'd added a section to the instrument parameter files and this was causing some changes in the instrument definition to show up in the system test comparisons between the generated \reference files. But it looks like you've disabled the instrument check in the system tests?

Indeed, I disabled checking the instrument as the new parameters were tripping the validation of system tests for `SANSILLAutoProcess`.  I am not sure I follow your comment here, as all of the added reference files (containing 'Logbook' in the name) are new files added for testing of the logbook generation for all supported instruments, and they do not replace previously present references.

File: instrument/D11B_Parameters.xml
Comment (DannyHindson) [reply instrument/D11B_Parameters.xml:29]: OK I see. So the Mantid parameter map functionality only permits\expects a single value. I guess you could embed the xml values list inside a single val attribute using something like:
```
<value val="&lt;logvalue=&quot;wavelength&quot; path=&quot;/entry0/D7/monochromator/wavelength&quot;\&gt;"\>
```
...but this is perhaps less user friendly than the approach you've taken. So what you have done looks fine.

Have you got the divide operator to work OK? It seems like it will be difficult for the code to correctly identify the "//" string inside a formula when the two paths also start with a "/" character. Is this the syntax you would use for a divide?

/entry0/D7/POL/actual_state///entry0/D7/POL/actual_stateB1B2

File: Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py
Comment (DannyHindson) [reply Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py:65]: Maybe you've not finished work on this yet, but I don't think this unit test has been updated? I can see you've updated the column header for entries that use a formula thanks.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (DannyHindson) [inline Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:54]: Could you validate that the length is equal to 2 here? Unless there are some use cases I've not thought about where you accept multiple ranges. Seems like you just read the [0] and [-1] elements though

File: instrument/D11B_Parameters.xml
Comment (darominski) [reply instrument/D11B_Parameters.xml:29]: Yes indeed, it seems that only one value can be assigned to a parameter. It's good to know there exist a workaround should one require it.

I have tested the behaviour of the operations and regexes in a separate small-scale script  to verify different input types and operations. I can send it your way if you would like to play with it outside of Mantid for more flexibility. The division is going to be performed as you showed, when 3 subsequent slashes are present, which is not the most appealing solution but it does work.

File: Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py
Comment (darominski) [reply Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py:65]: My apologies, after adding this correction to the doc test, I missed changing the unit test. I've just corrected it.

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:54]: Unfortunately, in that case all range specifiers similar to those used in `MultiFileProperty` would not be allowed. The implemented solution is not the most elegant, as it creates a list based on the input. If two elements are provided, that naturally gives the limits for the numor range. If a range is specified, e.g. through "360:365" or equally "360-365", a list of `int` containing all values ranging from 360 to 365, inclusive, is created and the code takes the first and last values as the numor range. Such a list is going to be longer than the desired 2 element, and during validation, the property will return a list, not the user input, so I am not sure here how to validate against it. A solution could be to provide more comprehensive tooltip.

File: instrument/D11B_Parameters.xml
Comment (DannyHindson) [reply instrument/D11B_Parameters.xml:29]: That's fine, no need to send me the script. I wasn't sure if the code would use the 1st\2nd slash or the 2nd\3rd in a group of three slashes as the divide operator but I've just tried it and it looks like you've sorted all of this out.
You could possibly add a space around the "//" when building up the header at ~L190 of GenerateLogBook.py so it's clear that the operator is unrelated to the slashes in the path but this is "nice to have". Perhaps let's agree on whether there are any other changes first...

File: Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py
Comment (DannyHindson) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:54]: If the ranges are being expanded out for you by the C++ code then I think you can just use the list in _prepare_file_array rather than extracting the [0] and [-1] elements ie
```
            try:
                numor = int(os.path.splitext(file[instrument_name_len:])[0])
                if numor in self._numor_range:
                    file_list.append(os.path.splitext(file)[0])
            except (ValueError, OverflowError):
                self.log().debug("File {} cannot be cast into an integer numor".format(file))
                continue
```
I just tried this and seemed to work. I saw a problem with filenames containing strings like 394458_394882 which seems to get successfully converted to an integer so that's why I added the OverflowError handling.

I think this would mean the "," no longer indicates a range and instead just indicates a list. But that seems OK and consistent with the "," elsewhere in Mantid?

Comment (darominski) [reply Framework/PythonInterface/plugins/algorithms/GenerateLogbook.py:54]: Thank you for this suggestion, this is pretty neat. In this form, a list with all desired numors can be provided and the file names are going to be verified against it. The user can also provide a range of numors that will be turned into a list which can be used to validate against. A slight modification of the behaviour when the `NumorRange` property is not specified is required, but that's all. Now, this modification can support more robust workflow.

**Comment by DannyHindson** [comment]
@darominski - this PR looks good to me now but some of the jobs seem to have got stuck\disappeared. Perhaps it's worth pushing an empty commit to retrigger the jobs? I'll approve once it's passed most of the jobs.
Also it's up to you but you may also want to squash some of the commits with a rebase before it gets merged

**Review by DannyHindson — APPROVED** [review]
This looks good now thanks

**Pull request merged**
